### PR TITLE
MINIFICPP-1225 - Fix flaky HTTP tests.

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/http-curl/tests/HTTPHandlers.h [hunk 1]
```diff
@@ -206,25 +207,36 @@ class FlowFileResponder : public CivetHandler {
       minifi::io::CivetStream civet_stream(conn);
       minifi::io::CRCStream < minifi::io::CivetStream > stream(&civet_stream);
       uint32_t num_attributes;
+      int read;
       uint64_t total_size = 0;
-      total_size += stream.read(num_attributes);
+      read = stream.read(num_attributes);
+      if(!isServerRunning())return false;
```
Comment (arpadboda) [inline extensions/http-curl/tests/HTTPHandlers.h:213]: Why?

I would expect not even trying to read the stream when the server is not running.

File: extensions/http-curl/client/HTTPStream.h [hunk 2]
```diff
@@ -155,13 +155,13 @@ class HttpStream : public io::BaseStream {
...
-   * type name.
+   * Populates the vector using the provided type name.
+   * @param buf output buffer
    * @param t incoming object
-   * @returns vector.
+   * @returns number of bytes read.
    */
   template<typename T>
-  std::vector<uint8_t> readBuffer(const T&);
+  int readBuffer(std::vector<uint8_t>&, const T&);
```
Comment (arpadboda) [inline extensions/http-curl/client/HTTPStream.h:164]: I didn't check the code of BaseStream, just to confirm: the original function was declared at this level and it's not an override of a base func?

Comment (szaszm) [reply extensions/http-curl/client/HTTPStream.h:164]: Correct, it's an inline non-virtual member function that happens to be defined in many stream classes the same way. A lot of code duplication, but works.

**Comment by arpadboda** [comment]
Looking at CI logs, there is something left:

https://travis-ci.org/github/apache/nifi-minifi-cpp/jobs/687464775

The last debug line before the failure of the testcase is pretty talkative:

```
[2020-05-15 14:46:30.368] [org::apache::nifi::minifi::processors::ListenHTTP::Handler] [debug] ListenHTTP handling POST request of length -1
```

ListenHTTP also uses civetweb to achieve listening, so we might have the same issue there as we have with tests.

File: extensions/http-curl/tests/HTTPHandlers.h (hunk 1 above)
Comment (adamdebreceni) [reply extensions/http-curl/tests/HTTPHandlers.h:213]: apparently they do not segfault, but simply indicate error in their return value [mg_read](https://github.com/civetweb/civetweb/blob/master/docs/api/mg_read.md)

File: libminifi/include/io/CRCStream.h [hunk 3]
```diff
@@ -145,17 +145,15 @@ class CRCStream : public BaseStream {
...
+   * Populates the vector using the provided type name.
+   * @param buf output buffer
    * @param t incoming object
-   * @returns vector.
+   * @returns number of bytes read.
    */
   template<typename K>
-  std::vector<uint8_t> readBuffer(const K& t) {
-    std::vector<uint8_t> buf;
+  int readBuffer(std::vector<uint8_t>& buf, const K& t) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h:154]: the problem is that every other read*/write* method confirms to the "return the number of read/wrote bytes or a negative number on error", throwing would deviate from that significantly, do we have a list of dependant projects, or a guide on what we consider part of the api?

[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/io/CRCStream.h:154]: In theory everything in libminifi headers is API. In practice we sometimes break non-core APIs if it fixes important issues or only breaks arcane uses. IIRC the stream APIs are considered "core" (important for 3rd party extension development), but @arpadboda will be able to confirm/correct me.

[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/io/CRCStream.h:154]: This leaves the option of a new overload with proper error handling. If existing 3rd party extensions happened to use the old overload, it should work similarly, even is without error checking.

If you have a better idea, feel free to propose it. I have a few more as well, but they also highly deviate from the style of the codebase with limited benefit.

[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/io/CRCStream.h:167]: readded the previous versions to libminifi

**Comment by arpadboda** [comment]
CI jobs green, executed the tests parallel on my server 3 times, still all green. Nice work!

**Pull request closed without merging**